File: screener/backtester/slippage.py
"""Slippage models for fill-price adjustment.

Each model exposes ``adverse_fraction(side, shares, adv, sigma_daily) -> float``
returning a non-negative fraction by which the reference price is widened
against the trader. ``apply_slippage`` multiplies that through a reference
price, with direction handled for buys vs sells.

Models:
  * ``FixedBpsSlippage`` — constant basis-point adverse fill (legacy behaviour).
  * ``HalfSpreadSlippage`` — quoted half-spread charged on every fill.
  * ``VolumeImpactSlippage`` — Almgren-Chriss sqrt-law impact:
    ``k * sigma_daily * sqrt(shares / adv_shares)``.
  * ``CompositeSlippage`` — sums adverse fractions from component models.
"""

from __future__ import annotations

import math
from typing import Literal, Protocol, runtime_checkable

from pydantic import BaseModel, ConfigDict, Field


Side = Literal["buy", "sell"]


@runtime_checkable
class SlippageModel(Protocol):
    def adverse_fraction(
        self, side: Side, shares: float, adv: float, sigma_daily: float
    ) -> float:
        """Return the adverse price adjustment as a non-negative fraction."""


def apply_slippage(
    model: SlippageModel,
    reference_price: float,
    side: Side,
    shares: float = 0.0,
    adv: float = 0.0,
    sigma_daily: float = 0.0,
) -> float:
    frac = float(model.adverse_fraction(side, shares, adv, sigma_daily))
    if frac < 0.0:
        frac = 0.0
    if side == "buy":
        return reference_price * (1.0 + frac)
    return reference_price * (1.0 - frac)


class FixedBpsSlippage(BaseModel):
    model_config = ConfigDict(frozen=True)

    bps: float = 0.0

    def adverse_fraction(
        self, side: Side, shares: float, adv: float, sigma_daily: float
    ) -> float:
        return self.bps / 10_000.0


class HalfSpreadSlippage(BaseModel):
    model_config = ConfigDict(frozen=True)

    half_spread_bps: float = 0.0

    def adverse_fraction(
        self, side: Side, shares: float, adv: float, sigma_daily: float
    ) -> float:
        return self.half_spread_bps / 10_000.0


class VolumeImpactSlippage(BaseModel):
    """Almgren-Chriss square-root-law market impact.

    ``adv`` is expected in shares (not dollars). When ADV is unknown or zero
    the model returns 0 rather than raising, so the caller can fall through
    to other components in a composite.
    """

    model_config = ConfigDict(frozen=True)

    k: float = 0.1

    def adverse_fraction(
        self, side: Side, shares: float, adv: float, sigma_daily: float
    ) -> float:
        if adv <= 0.0 or shares <= 0.0 or sigma_daily <= 0.0:
            return 0.0
        return self.k * sigma_daily * math.sqrt(shares / adv)


class CompositeSlippage(BaseModel):
    model_config = ConfigDict(frozen=True, arbitrary_types_allowed=True)

    models: tuple[SlippageModel, ...] = Field(default_factory=tuple)

    def adverse_fraction(
        self, side: Side, shares: float, adv: float, sigma_daily: float
    ) -> float:
        total = 0.0
        for m in self.models:
            total += float(m.adverse_fraction(side, shares, adv, sigma_daily))
        return total

File: screener-rs/src/backtester/slippage.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Side {
    Buy,
    Sell,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "kebab-case")]
pub enum SlippageModel {
    Fixed {
        bps: f64,
    },
    HalfSpread {
        half_spread_bps: f64,
    },
    VolumeImpact {
        k: f64,
    },
    Composite {
        fixed_bps: f64,
        half_spread_bps: f64,
        vol_impact_k: f64,
    },
}

impl Default for SlippageModel {
    fn default() -> Self {
        Self::Fixed { bps: 0.0 }
    }
}

impl SlippageModel {
    pub fn adverse_fraction(&self, shares: f64, adv: f64, sigma_daily: f64) -> f64 {
        match *self {
            SlippageModel::Fixed { bps } => bps / 10_000.0,
            SlippageModel::HalfSpread { half_spread_bps } => half_spread_bps / 10_000.0,
            SlippageModel::VolumeImpact { k } => volume_impact(k, shares, adv, sigma_daily),
            SlippageModel::Composite {
                fixed_bps,
                half_spread_bps,
                vol_impact_k,
            } => {
                fixed_bps / 10_000.0
                    + half_spread_bps / 10_000.0
                    + volume_impact(vol_impact_k, shares, adv, sigma_daily)
            }
        }
    }
}

fn volume_impact(k: f64, shares: f64, adv: f64, sigma_daily: f64) -> f64 {
    if adv <= 0.0 || shares <= 0.0 || sigma_daily <= 0.0 {
        return 0.0;
    }
    k * sigma_daily * (shares / adv).sqrt()
}

pub fn apply_slippage(
    model: &SlippageModel,
    reference_price: f64,
    side: Side,
    shares: f64,
    adv: f64,
    sigma_daily: f64,
) -> f64 {
    let frac = model.adverse_fraction(shares, adv, sigma_daily).max(0.0);
    match side {
        Side::Buy => reference_price * (1.0 + frac),
        Side::Sell => reference_price * (1.0 - frac),
    }
}
